app/src/main/java/com/futurumgame/base/unlockables/UnlockableCollection.java

```diff
@@ -77,16 +100,116 @@ public void updateNewUnlock(Unlockable unlocked, WareHouse wareHouse) {
+        for(Unlockable buyable: CollectionHelper.whereNotNull(buyables)){
+            if(buyableUnlockables.contains(buyable)){
+                continue;
+            }
+            buyableUnlockables.add(buyable);
+        }
+        if(!buyables.isEmpty() || CollectionHelper.sequenceEquals(personalUnlockables(), boughtUnlockables)) {
+            return;
+        }
+        correctFileCorruption();
+    }
+
+    @Override
+    public IParseRule<Unlockable> getParseRule() {
+        return ParseRule;
+    }
+
+    @Override
+    public byte[] provideData() {
+        String buyables = CollectionHelper.toString(ParseRule::getParsingValue,Separator.DefaultSeparator, buyableUnlockables);
+        String bought = CollectionHelper.toString(ParseRule::getParsingValue,Separator.DefaultSeparator, boughtUnlockables);
+        return DataEncoding.UTF8.encode(StringUtil.combine(Separator.CollectionSeparator, buyables, bought));
+    }
+
+    private void addInitialUnlockable() {
+        buyableUnlockables.add(InitialUnlockables.get(targetUnlockables));
+    }
+
+    private void correctFileCorruption() {
+        if(boughtUnlockables.isEmpty()) {
+            Logger.i(getClass(), "starting from scratch");
+            addInitialUnlockable();
+            return;
+        }
+        //TODO fix other possible corruption states, currently no further known state
+    }
+
+    private HashSet<Unlockable> personalUnlockables(){
+        return AllUnlockables.get(targetUnlockables);
+    }
+
+    private Unlockable parse(String s) {
+        ParseResult<Unlockable> result = ParseRule.next(s);
+        if (result.parseSuccess()) {
+            return result.getResult();
+        }
+        return null;
+    }
+
+    private void simulateUnlocks(String[] resources) {
+        LinkedList<Unlockable> bought = CollectionHelper.select(resources, this::parse);
+        buyableUnlockables.addAll(CollectionHelper.whereNotNull(bought));
+        LinkedList<Unlockable> temp = new LinkedList<>(buyableUnlockables);
+        temp.forEach(Unlockable::unlock);
+    }
+
+    public static UnlockableCollection createResourceUnlockables() {
+        return new UnlockableCollection(ResourceUnlockable.class);
+    }
+
+    public static UnlockableCollection createResearchUnlockables() {
+        return new UnlockableCollection(ResearchUnlockable.class);
+    }
+
+    public static Units getDefaultResourceCapacity(int id) {
+        if(id == Water.ID){
+            return WareHouse.WaterBaseCapacity;
+        }
+        HashSet<Unlockable> resourceUnlockables = AllUnlockables.get(ResourceUnlockable.class);
+        Map<Integer, Unlockable> unlockablesByResourceID = CollectionHelper.toMap(resourceUnlockables, u->((ResourceUnlockable) u).getResourceID());
+        ResourceUnlockable unlockable = (ResourceUnlockable) unlockablesByResourceID.get(id);
+        return unlockable.getStartCap();
+    }
+
+    //Not Cleancode!!!! TODO remove dependecy on init order
```

Comment:
Then do it I guess?

Reply:
will be done upon switch to c# and godot